server/src/main/java/org/elasticsearch/cluster/coordination/StableMasterHealthIndicatorService.java

```diff
@@ -0,0 +1,433 @@
+     * in the last 30 minutes). This method attemtps to use the master history from a remote node to confirm what we are seeing locally.
+     * If the information from the remote node confirms that the master history has been unstable, a YELLOW status is returned. If the
+     * information from the remote node shows that the master history has been stable, then we assume that the problem is with this node
+     * and a GREEN status is returned (the problems with this node will be covered in a different health indicator). If there had been
+     * problems fetching the remote master history, the exception seen will be included in the details of the result.
+     * @param localMasterHistory The master history as seen from the local machine
+     * @param explain Whether to calculate and include the details in the result
+     * @return The HealthIndicatorResult for the given localMasterHistory
+     */
+    private HealthIndicatorResult calculateWhenMasterHasFlappedNull(MasterHistory localMasterHistory, boolean explain) {
+        DiscoveryNode master = localMasterHistory.getMostRecentNonNullMaster();
+        boolean localNodeIsMaster = clusterService.localNode().equals(master);
+        List<DiscoveryNode> remoteHistory;
+        Exception remoteHistoryException = null;
+        if (localNodeIsMaster) {
+            remoteHistory = null; // We don't need to fetch the remote master's history if we are that remote master
+        } else {
+            try {
+                remoteHistory = masterHistoryService.getRemoteMasterHistory();
+            } catch (Exception e) {
+                remoteHistory = null;
+                remoteHistoryException = e;
+            }
+        }
+        /*
+         * If the local node is master, then we have a confirmed problem (since we now know that from this node's point of view the
+         * master is unstable).
+         * If the local node is not master but the remote history is null then we have a problem (since from this node's point of view the
+         * master is unstable, and we were unable to get the master's own view of its history). It could just be a short-lived problem
+         * though if the remote history has not arrived yet.
+         * If the local node is not master and the master history from the master itself reports that the master has gone null repeatedly
+         *  or changed identity repeatedly, then we have a problem (the master has confirmed what the local node saw).
+         */
+        boolean masterConfirmedUnstable = localNodeIsMaster
+            || remoteHistory == null
+            || MasterHistory.hasMasterGoneNullAtLeastNTimes(remoteHistory, acceptableNullTransitions + 1)
+            || MasterHistory.getNumberOfMasterIdentityChanges(remoteHistory) > acceptableIdentityChanges;
+        if (masterConfirmedUnstable) {
+            if (localNodeIsMaster == false && remoteHistory == null) {
```

Comment:
if remoteHistory is null but there's an in-flight request to get it I believe the status should still be GREEN?